lucille-plugins/lucille-pinecone/src/main/java/com/kmwllc/lucille/pinecone/stage/EmitDocsToDeleteByPrefix.java

```diff
@@ -0,0 +1,149 @@
+ * - deletionMarkerField (String) : name of the field which indicates the document is marked for deletion
+ * - deletionMarkerFieldValue (String) : value of the field which indicates that the document is marked for deletion
+ *   - for stage to work, deletionMarkerField and deletionMarkerFieldValue must be the same value provided indexer configs
+ */
+public class EmitDocsToDeleteByPrefix extends Stage {
+
+  private Pinecone client;
+  private final String indexName;
+  private Index index;
+  private final Map<String, Object> namespaces;
+  private final boolean dropOriginal;
+  private final String addPrefix;
+  private final String deletionMarkerField;
+  private final String deletionMarkerFieldValue;
+
+  public EmitDocsToDeleteByPrefix(Config config) {
+    super(config, new StageSpec()
+        .withOptionalProperties("dropOriginal", "addPrefix")
+        .withOptionalParents("namespaces")
+        .withRequiredProperties("apiKey", "deletionMarkerField", "deletionMarkerFieldValue", "index"));
+    this.indexName = config.getString("index");
+    this.namespaces = config.hasPath("namespaces") ? config.getConfig("namespaces").root().unwrapped() : null;
+    this.dropOriginal = config.hasPath("dropOriginal") ? config.getBoolean("dropOriginal") : false;
+    this.addPrefix = config.hasPath("addPrefix") ? config.getString("addPrefix") : "";
+    this.deletionMarkerField = config.getString("deletionMarkerField");
+    this.deletionMarkerFieldValue = config.getString("deletionMarkerFieldValue");
+  }
+
+  @Override
+  public void start() throws StageException {
+    try {
+      client = PineconeManager.getClient(config.getString("apiKey"));
+      index = client.getIndexConnection(indexName);
+    } catch (Exception e) {
+      throw new StageException("Unable to get client or index connection", e);
+    }
+
+    StateEnum state = client.describeIndex(indexName).getStatus().getState();
+    boolean isStable = state != StateEnum.INITIALIZING &&
+        state != StateEnum.INITIALIZATIONFAILED &&
+        state != StateEnum.TERMINATING;
+    if (!isStable) {
+      throw new StageException("Index " + indexName + " is not stable");
+    }
+  }
+
+  @Override
+  public Iterator<Document> processDocument(Document doc) throws StageException {
+    // check if document has been set for deletion and if not skip current document
+    if (!isMarkedForDeletion(doc)) {
+      return null;
+    }
+
+    List<String> idsToDelete = new ArrayList<>();
+    // collect all documents that meet this prefix in all namespaces given
+    if (namespaces != null) {
+      for (Map.Entry<String, Object> entry : namespaces.entrySet()) {
+        idsToDelete.addAll(getVectorsByPrefix(doc, addPrefix, entry.getKey()));
+      }
+    } else {
+      idsToDelete.addAll(getVectorsByPrefix(doc, addPrefix, PineconeManager.getDefaultNamespace()));
+    }
+
+    // drop original document if set to true
+    doc.setDropped(dropOriginal);
```

Comment:
very minor comment but I'd suggest slightly more verbose

`if (dropOriginal) { doc.setDropped(true);}`

because it's a bit more specific and explicit about the intent. If there were other code somewhere above that marked the doc as dropped (I know there isn't, but there could be) then doc.setDropped(dropOriginal) would actually un-drop it if dropOriginal is false.